src/daemon/auth/providers/openid_connect/util.rs

```diff
@@ -0,0 +1,226 @@
+use openidconnect::core::{
+    CoreAuthDisplay, CoreAuthPrompt, CoreClaimName, CoreClaimType, CoreClientAuthMethod, CoreErrorResponseType,
+    CoreGenderClaim, CoreGrantType, CoreJsonWebKey, CoreJsonWebKeyType, CoreJsonWebKeyUse,
+    CoreJweContentEncryptionAlgorithm, CoreJweKeyManagementAlgorithm, CoreJwsSigningAlgorithm, CoreResponseMode,
+    CoreResponseType, CoreSubjectIdentifierType, CoreTokenType,
+};
+use openidconnect::{
+    AdditionalClaims, AdditionalProviderMetadata, Client, ExtraTokenFields, IdTokenClaims, IdTokenFields,
+    ProviderMetadata, StandardErrorResponse, StandardTokenResponse, UserInfoClaims,
+};
+
+use crate::commons::error::Error;
+use crate::commons::KrillResult;
+
+// -----------------------------------------------------------------------------
+// Swap out the openidconnect crate types EmptyAdditionalClaims and
+// EmptyExtraTokenFields with our own types which are flexible enough to handle
+// deserialization of any possible (but known in advance) claim/token fields
+// that the OpenID Connect provider might include in a response to us. Otherwise
+// the default openidconnect crate implementation drops such response fields
+// during deserialization preventing us from extracting them (e.g. to learn
+// which Krill role the identity provider logged in user should have).
+// -----------------------------------------------------------------------------
+
+// Define additional claims that we might receive from the OpenID Connect
+// provider as an arbitrary JSON hierarchy as we cannot know at compile time
+// which claim name we should expect to look for from a given customers provider
+// rather they must tell us that via runtime configuration. If we were instead
+// to, for example, define a "role" member field of our custom additional claims
+// struct, serde_json would fail to deserialize it if the the field is not
+// present or not structured as expected. Using this approach we can inspect the
+// structure when we receive it from the provider.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CustomerDefinedAdditionalClaims(serde_json::Value);
+impl AdditionalClaims for CustomerDefinedAdditionalClaims {}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CustomerDefinedExtraTokenFields(serde_json::Value);
```

Comment:
The oauth2-rs and openidconnect-rs crates do this slightly differently, maybe their way would be better?
E.g. see [here](https://github.com/ramosbugs/oauth2-rs/blob/176fbdd4ae97868f20757690329fe05b71a23652/examples/google_devicecode.rs#L26) and [here](https://github.com/ramosbugs/openidconnect-rs/blob/f4a7abef527a22a905521a198a98b6f1b435c422/src/user_info.rs#L539). Rather than `(serde_json::Value)` they use `(HashMap<String, serde_json::Value)`.